### DISCUSSION TITLE
Strange behaviour of org.eclipse.equinox.app when running under different os

### DISCUSSION THREAD

**Discussion by laeubi** [eclipse-equinox/equinox#242] [Q&A]
Today I debugged a problem where starting an application failed under different configurations, when one runs the application under Windows/Linux everything works, but under MacOS it failed with:

> Unable to acquire application service. Ensure that the **org.eclipse.core.runtime** bundle is resolved and started (see config.ini).

In contrast to what is indicated by the error message, it seems that `org.eclipse.equinox.app` must be explicitly started, and I found that it was in starting (state = 8) state instead. 

```
4 = resolved
8 = starting
32 = active
```

**List of bundles that was not working:**
```
32 | org.eclipse.osgi_3.18.200.v20221116-1324 [0]
8  | org.eclipse.core.contenttype_3.8.200.v20220817-1539 [1]
8  | org.eclipse.core.jobs_3.13.200.v20221102-1024 [2]
8  | org.eclipse.core.runtime_3.26.100.v20221021-0005 [3]
8  | org.eclipse.equinox.app_1.6.200.v20220720-2012 [4]
8  | org.eclipse.equinox.common_3.17.0.v20221108-1156 [5]
4  | org.eclipse.equinox.launcher_1.6.400.v20210924-0641 [6]
8  | org.eclipse.equinox.preferences_3.10.100.v20220710-1223 [7]
8  | org.eclipse.equinox.registry_3.11.200.v20220817-1601 [8]
.. some other bundles containing the application all in state resolved ..
4  | org.osgi.service.prefs_1.1.2.202109301733 [13]
```

**List of bundles where it was working**
```
32 | org.eclipse.osgi_3.18.200.v20221116-1324 [0]
8  | org.eclipse.core.contenttype_3.8.200.v20220817-1539 [1]
8  | org.eclipse.core.jobs_3.13.200.v20221102-1024 [2]
8  | org.eclipse.core.runtime_3.26.100.v20221021-0005 [3]
32 | org.eclipse.equinox.app_1.6.200.v20220720-2012 [4]
32 | org.eclipse.equinox.common_3.17.0.v20221108-1156 [5]
4  | org.eclipse.equinox.launcher_1.6.400.v20210924-0641 [6]
8  | org.eclipse.equinox.preferences_3.10.100.v20220710-1223 [7]
32 | org.eclipse.equinox.registry_3.11.200.v20220817-1601 [8]
.. some other bundles containing the application all in state resolved ..
4  | org.osgi.service.prefs_1.1.2.202109301733 [13]
```

So there are actually two questions:

1. can the properties influence how/if bundles are started? I first thought that maybe a platform specific fragment can make a difference but there are actual no fragment is resolved
2. should not state 8 (starting) be enough for the application to start up?

**Comment by tjwatson** [comment]
That error message is really out of date ` (see config.ini)`.  Since the introduction of p2 the default `config.ini` only has:
`osgi.bundles=reference\:file\:org.eclipse.equinox.simpleconfigurator_<version>.jar@1\:start`

Then simple configurator is responsible for installing/starting all other bundles according to the `org.eclipse.equinox.simpleconfigurator/bundles.info`

I would expect the following bundles to all be marked for eager activation (also note the start-levels):

```
org.apache.felix.scr,2.2.6,plugins/org.apache.felix.scr_2.2.6.jar,2,true
org.eclipse.core.runtime,3.27.0.v20230317-0901,plugins/org.eclipse.core.runtime_3.27.0.v20230317-0901.jar,4,true
org.eclipse.equinox.common,3.17.100.v20230202-1341,plugins/org.eclipse.equinox.common_3.17.100.v20230202-1341.jar,2,true
org.eclipse.equinox.event,1.6.200.v20230120-0604,plugins/org.eclipse.equinox.event_1.6.200.v20230120-0604.jar,2,true
org.eclipse.equinox.p2.reconciler.dropins,1.4.200.v20220911-2007,plugins/org.eclipse.equinox.p2.reconciler.dropins_1.4.200.v20220911-2007.jar,4,true
org.eclipse.equinox.simpleconfigurator,1.4.200.v20221111-1340,plugins/org.eclipse.equinox.simpleconfigurator_1.4.200.v20221111-1340.jar,1,true
org.eclipse.osgi,3.18.300.v20230220-1352,plugins/org.eclipse.osgi_3.18.300.v20230220-1352.jar,-1,true
```

I'm unsure how you have any "working" scenario where `org.eclipse.core.runtime` is not eagerly activated.

**Reply by laeubi** [reply]
I'm not using simple configuration but only `bundle install` + `bundle start`, here is an example where I try to run the ApiApplication:

https://github.com/eclipse-platform/eclipse.platform/actions/runs/4503893184/jobs/7927673758?pr=369

in the error output you see:

```
Error:  Execute the Api application failed! API Tools failure! -> [Help 1]
Error:  Execute the Api application failed! API Tools failure! -> [Help 1]
Error:  Execute the Api application failed! API Tools failure! -> [Help 1]
Error:  Execute the Api application failed! Unable to acquire application service. Ensure that the org.eclipse.core.runtime bundle is resolved and started (see config.ini). -> [Help 1]
Error:  Execute the Api application failed! Unable to acquire application service. Ensure that the org.eclipse.core.runtime bundle is resolved and started (see config.ini). -> [Help 1]
Error:  Execute the Api application failed! Unable to acquire application service. Ensure that the org.eclipse.core.runtime bundle is resolved and started (see config.ini). -> [Help 1]
```

So some launches fail with the "Unable to acquire application service", while others fail with "API Tools failure" (what is expected and shows the application was successfully launched) of course all launches using the same configuration here!
On the windows + linux nodes all runs  succeeds ... of cause there is also the same code running.

I have now adjusted tycho to always start the app bundle:
- https://github.com/eclipse-tycho/tycho/pull/2280/files

but would have expected that it either **always** fails or **never fails* ...

**Comment by tjwatson** [comment]
I'm not sure why one platform would fail and the other not.  Must be some ordering issue.  Regardless, I'm still not clear on what you are doing.  The most consistent way to get things to work reliably would be to start all the bundles in the same way simple configurator would with the platform installation.

That is, first set all the start-levels of all the bundles to the appropriate levels.  By default simpleconfigurator sets everything to start-level 4.  The exception is the following list:

```
org.apache.felix.scr,2.2.6,plugins/org.apache.felix.scr_2.2.6.jar,2,true
org.eclipse.core.runtime,3.27.0.v20230317-0901,plugins/org.eclipse.core.runtime_3.27.0.v20230317-0901.jar,4,true
org.eclipse.equinox.common,3.17.100.v20230202-1341,plugins/org.eclipse.equinox.common_3.17.100.v20230202-1341.jar,2,true
org.eclipse.equinox.event,1.6.200.v20230120-0604,plugins/org.eclipse.equinox.event_1.6.200.v20230120-0604.jar,2,true
org.eclipse.equinox.p2.reconciler.dropins,1.4.200.v20220911-2007,plugins/org.eclipse.equinox.p2.reconciler.dropins_1.4.200.v20220911-2007.jar,4,true
org.eclipse.equinox.simpleconfigurator,1.4.200.v20221111-1340,plugins/org.eclipse.equinox.simpleconfigurator_1.4.200.v20221111-1340.jar,1,true
org.eclipse.osgi,3.18.300.v20230220-1352,plugins/org.eclipse.osgi_3.18.300.v20230220-1352.jar,-1,true
```

This indicates the start-level and that these are `eagerly` started.  For example, `org.eclipse.equinox.common` has start-level `2`.  All other bundles would have start-level 4 and would would be lazy started.
